software/contrib/master_clock.py

```diff
@@ -0,0 +1,340 @@
+- Vary BPM by sending CV into the analog input
+- Set the division to zero for an output, this will cause the output to randomly go from high (+5V) to low (0V)
+
+Demo video: TBC
+
+digital_in: (optional) Reset step count on rising edge
+analog_in: (optional) Adjust BPM
+
+knob_1: Screen 2: Adjust BPM. Screen 3: Select output to edit 
+knob_2: Screen 2: Adjust Pulse width. Screen 3: Adjust division of selected output 
+
+button_1: Start / Stop
+button_2: Short Press (<500ms): Cycle through screens. Long Press (>500ms): Enter config mode
+
+Defaults:
+output_1: clock / 1
+output_2: clock / 2
+output_3: clock / 4
+output_4: clock / 8
+output_5: clock / 16
+output_6: clock / 32
+
+Known Issues:
+- If playback is restarted while screen 2 is in config mode, playback will be slightly irratic
+
+'''
+
+class MasterClockInner(EuroPiScript):
+    def __init__(self):
+        # Overclock the Pico for improved performance.
+        machine.freq(250_000_000)
+        self.step = 1
+        self.completedCycles = 0
+        self.running = True
+        self.CLOCKS_PER_QUARTER_NOTE = 4
+        self.MIN_BPM = 20  # Successfully calibrated to >= 20 and <= 240 BPM
+        self.MAX_BPM = 240
+        self.MIN_PULSE_WIDTH = 8
+        self.resetTimeout = 2000
+        self.previousStepTime = 0
+        self.minAnalogInputVoltage = 0.9
+        self.screen = 1
+        self.configMode = False
+        self.screenIndicator = ['.  ', ' . ', '  .' ]
+
+        self.MAX_DIVISION = 128
+        self.clockDivisions = []
+        for n in range(self.MAX_DIVISION+1):
+            self.clockDivisions.append(n)
+        
+        # When enabled, set msDriftCompensation to 30
+        # When disabled, set msDriftCompensation to 28
+        self.DEBUG = False
+
+        # In testing a 28ms drift was found at all tempos
+        # Adding this offset brings the BPM and pulse width back into a reasonable tolerance
+        self.msDriftCompensation = 28
+
+        # Vars to drive UI on screen3
+        self.markerPositions = [ [0, 1], [0, 11], [0, 21], [52, 1], [52, 11], [52, 21]]
+        self.activeOption = 1
+        self.previousselectedDivision = ''
+
+        # Get working vars
+        self.loadState()
+        self.calcSleepTime()
+        self.getPulseWidth()
+
+        # Get asyncio event loop object
+        self.el = asyncio.get_event_loop()
+        self.el.run_forever()
+
+        # Starts/Stops the master clock
+        @b1.handler_falling
+        def StartStop():
+            self.running = not self.running
+
+        # Cycle screen and toggle config mode
+        @b2.handler_falling
+        def cycleScreen():
+            if ticks_diff(ticks_ms(), b2.last_pressed()) > 500 and ticks_diff(ticks_ms(), b2.last_pressed()) < 4000:
+                self.configMode = not self.configMode
+                self.running = False
+            else:
+                # Turn off config mode to avoid current knob positions messing up other settings on the next screen
```

Comment:
[optional] I recently committed the `LockableKnob` class which may be helpful to you here.

See `https://allen-synthesis.github.io/EuroPi/generated/experimental.knobs.html#experimental.knobs.LockableKnob`

As well as this note about the experimental package that it lives in:  https://allen-synthesis.github.io/EuroPi/generated/experimental.html

Reply:
I am going to hold off on this change for now until I hear back from others on the UI. UI changes may negate the need for this hack in the end anyway. Thanks for the info though.

Reply:
Implemented a slightly different lockable mechanism as I needed different locking rules for different modes